jolt-core/src/benches/sum_timer.rs

```diff
@@ -0,0 +1,107 @@
+}
+
+pub struct FlushGuard {
+    span_durations: Arc<Mutex<HashMap<String, u128>>>,
+    filter: Option<Vec<String>>,
+}
+
+impl FlushGuard {
+    fn print_header(&self) {
+        println!("===============================");
+        println!("||                            ||");
+        println!("||     Sum Timing Results     ||");
+        println!("||                            ||");
+        println!("===============================");
+        if let Some(filter) = &self.filter {
+            println!("===============================");
+            println!("||     Filter Details         ||");
+            println!("===============================");
+            for name in filter {
+                let mut padding = if name.len() > 28 { 0 } else { 28 - name.len() };
+                let left_padding = padding / 2;
+                let right_padding = padding - left_padding;
+                println!("||{:>padding_left$}{}{:>padding_right$}||", "", name, "", padding_left=left_padding, padding_right=right_padding);
+            }
+            println!("===============================");
+        }
+    }
+}
+
+impl Drop for FlushGuard {
+    fn drop(&mut self) {
+        self.print_header();
+        let durations = self.span_durations.lock().unwrap();
+        match &self.filter {
+            Some(filter) => {
+                for name in filter {
+                    if let Some(duration) = durations.get(name) {
+                        println!("'{}': {:?} ms", name, duration);
+                    }
+                }
+            }
+            None => {
+                for (name, duration) in durations.iter() {
+                    println!("'{}': {:?} ms", name, duration);
+                }
+            }
+        }
+    }
+
+}
+
+impl SumTimingLayer {
+    /// Creates a new SumTimingLayer with an optional `filter: Option<Vec<String>>` of named spans.
+    /// Also returns a FlushGuard. Prints sum details when the FlushGuard is dropped.
+    pub fn new(filter: Option<Vec<String>>) -> (Self, FlushGuard) {
+        let span_durations = Arc::new(Mutex::new(HashMap::new()));
+        let layer = SumTimingLayer {
+            span_durations: Arc::clone(&span_durations),
+        };
+        let guard = FlushGuard {
+            span_durations: span_durations,
+            filter,
+        };
+        (layer, guard)
+    }
+}
+
+impl<S> Layer<S> for SumTimingLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_enter(&self, id: &tracing::span::Id, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            span.extensions_mut().insert(Instant::now());
+        }
+    }
+
+    fn on_exit(&self, id: &tracing::span::Id, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let span_name = span.name().to_string();
+            if let Some(start_time) = span.extensions_mut().remove::<Instant>() {
+                let duration = start_time.elapsed().as_millis();
```

Comment:
nit: store duration in micros and nanos, but still print out the cumulative numbers in millis (with a couple decimal places of precision)
otherwise I think we'll be underestimating stuff like `eval_loop`, `binding`

Reply:
Yup, good call.